### PR TITLE
do some webhook job in daemon

### PR DESCRIPTION
**Pull request by hhyasdf**
Pull Request Description
---

### Describe how you did it
1. Do some webhook job in daemon.
2. Remove ip family checking for pod ip cache
3. Return error if specified ip family is invalid
4. IP family should also be picked up in webhook way by manager if webhook is down
5. Stateful allocated pod should always can be recreated whatever the ip family and network type they are changed to
6. Get specified network/network-type in webhook way if webhook is not working
7. Get specified subnet in webhook way if webhook is not working

### Special notes for reviews

### CODE REVIEW

File: pkg/daemon/server/handle.go [hunk 1]
```diff
@@ -110,6 +105,17 @@ func (cdh *cniDaemonHandler) handleAdd(req *restful.Request, resp *restful.Respo
 	backOffBase := 5 * time.Microsecond
 	retries := 11
 	ipFamily := ipamtypes.ParseIPFamilyFromString(pod.Annotations[constants.AnnotationIPFamily])
+	handledByWebhook := globalutils.ParseBoolOrDefault(pod.Annotations[constants.AnnotationHandledByWebhook], false)
+
+	if !handledByWebhook {
+		_, _, _, ipFamilyStr, err := webhookutils.ParseNetworkConfigOfPodByPriority(context.TODO(), cdh.mgrAPIReader, pod)
```
Comment (mars1024) [inline pkg/daemon/server/handle.go:111]: ipFamilyStr maybe empty here, so we have to defaulting it if empty

Comment (hhyasdf) [reply pkg/daemon/server/handle.go:111]: done

File: pkg/webhook/mutating/pod.go [hunk 2]
```diff
@@ -133,6 +115,8 @@ func PodCreateMutation(ctx context.Context, req *admission.Request, handler *Han
 			patchSelectorToPod(pod, map[string]string{
 				constants.LabelDualStackAddressQuota: constants.QuotaNonEmpty,
 			})
+		default:
```
Comment (mars1024) [inline pkg/webhook/mutating/pod.go:118]: actually, ip family is validated in validating webhook before mutation, so `default` is unreachable.

File: pkg/webhook/utils/utils.go [hunk 3]
```diff
@@ -201,5 +180,69 @@ func ParseNetworkConfigOfPodByPriority(ctx context.Context, c client.Reader, pod
...
+		// ignore terminating ipInstance
+		if ipList.Items[i].DeletionTimestamp == nil {
+			networkName = ipList.Items[i].Spec.Network
+			ipFamilyCounter++
+		}
+	}
+
+	switch ipFamilyCounter {
+	case 1:
+		if networkingv1.IsIPv6IPInstance(&ipList.Items[0]) {
```
Comment (mars1024) [inline pkg/webhook/utils/utils.go:235]: should terminating IPInstance be ignored here?

File: pkg/webhook/utils/utils.go [hunk 4]
```diff
@@ -201,5 +180,69 @@ func ParseNetworkConfigOfPodByPriority(ctx context.Context, c client.Reader, pod
...
+		if err = c.Get(ctx, types.NamespacedName{Name: networkName}, network); err != nil {
+			err = fmt.Errorf("failed to get network %v: %v", networkName, err)
+			return
+		}
+
+		// specified network takes higher priority than network type defaulting, if no network type specified
+		// from pod, then network type should inherit from network type of specified network from pod
+		if len(networkTypeStr) == 0 {
+			networkType = ipamtypes.ParseNetworkTypeFromString(string(networkingv1.GetNetworkType(network)))
+		}
```
Comment (mars1024) [inline pkg/webhook/utils/utils.go:199]: maybe it's need to check the conflict between network-name and network-type

File: pkg/controllers/networking/pod_controller.go [hunk 5]
```diff
@@ -171,53 +171,60 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result
...
-		switch ipFamily {
-		case ipamtypes.IPv4, ipamtypes.IPv6, ipamtypes.DualStack:
-		default:
+		if networkStrFromWebhook, _, networkTypeFromWebhook,
+			ipFamilyFromWebhook, _, err = utils.ParseNetworkConfigOfPodByPriority(ctx, r, pod); err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to parse network config of pod: %v", err)
+		}
+
+		ipFamily = ipFamilyFromWebhook
+		if !ipamtypes.IsValidFamilyMode(ipFamily) {
```
Comment (mars1024) [inline pkg/controllers/networking/pod_controller.go:191]: ip family has been validated in `ParseNetworkConfigOfPodByPriority`, so this line is unreachable

File: pkg/daemon/server/handle.go [hunk 6]
```diff
@@ -110,6 +105,22 @@ func (cdh *cniDaemonHandler) handleAdd(req *restful.Request, resp *restful.Respo
...
+	if !handledByWebhook {
+		_, _, _, ipFamily, _, err = webhookutils.ParseNetworkConfigOfPodByPriority(context.TODO(), cdh.mgrAPIReader, pod)
+		if err != nil {
+			errMsg := fmt.Errorf("failed to parse network config of pod %v: %v", pod.Name, err)
+			cdh.errorWrapper(errMsg, http.StatusBadRequest, resp)
+			return
+		}
+	}
+
+	if !ipamtypes.IsValidFamilyMode(ipFamily) {
```
Comment (mars1024) [inline pkg/daemon/server/handle.go:119]: ip family validated in `PaseNetworkConfigOfPodByPriority`

File: pkg/webhook/mutating/pod.go (hunk 2 above)
Comment (hhyasdf) [reply pkg/webhook/mutating/pod.go]: done

File: pkg/webhook/utils/utils.go (hunk 3 above)
Comment (hhyasdf) [reply pkg/webhook/utils/utils.go]: done

File: pkg/webhook/utils/utils.go (hunk 4 above)
Comment (hhyasdf) [reply pkg/webhook/utils/utils.go:199]: done

File: pkg/controllers/networking/pod_controller.go (hunk 5 above)
Comment (hhyasdf) [reply pkg/controllers/networking/pod_controller.go]: done

File: pkg/daemon/server/handle.go (hunk 6 above)
Comment (hhyasdf) [reply pkg/daemon/server/handle.go]: done

**Review by mars1024 — APPROVED** [review]
/lgtm

**Pull request merged**